highway.tex

```diff
@@ -615,9 +630,207 @@ \subsection{Checkpoints}
+\end{enumerate}
+Together these imply that only votes satisfying Limited Naivety Criterion are sent and that if a vote is not considered to be delivered until it is removed from the verification buffer of the receiver, under the assumptions of condition 2 above, votes are being delivered within bounded time between honest validators, which is sufficient for our liveness strategy.
+
+The strategy that satisfies these conditions is relatively simple. An honest validator $V$ maintains a list $E_V$ of other validators for which $V$ knows of equivocating votes sent by that validator. When $V$ receives a new vote $\mu$ it is put in $V$'s verification buffer if either:
+\begin{itemize}
+\item The sender of $\mu$ is a known equivocator (i.e. $S(\mu) \in E_V$).
+\item The justification of $\mu$ contains some other vote already in $V$'s verification buffer.
+\end{itemize}
+Upon receiving such a vote, $V$ will send out to the other validators soliciting endorsements of it. When $V$ receives endorsements for any vote $\mu$ by a set of validators comprising at least $2/3$rds of the total validator weight (i.e. $V$ receives confirmation that $\mu$ is endorsed) and once $V$ has independently confirmed the validity of $\mu$ (perhaps by requesting from its sender other votes in its justification or proofs of endorsement of such votes), $V$ will remove $\mu$ and all votes in its justification from their verification buffer. Furthermore, upon receiving a request for endorsement for a vote $\mu$, $V$ will endorse this vote (forwarding the endorsement to the other validators) unless the sender of $\mu$ is in $E_V$.
+
+We now need to show that this strategy satisfies our requirements. First, note that since $V$ will only endorse votes by senders not known to be equivocators, $V$ will never endorse a pair of equivocating votes. This is because upon initial receipt of the second vote, $V$ will know the sender to be an equivocator and thus will not endorse. On the other hand, an honest validator will never equivocate and thus will never be on any other honest validator's list of equivocators. Thus, if endorsements are solicited for a vote $\mu$ sent by an honest validator $V$, all other honest validators will eventually endorse. Assuming that these honest validators comprise at least $2/3$rds of the total validator weight, this means that $\mu$ will eventually be endorsed. Furthermore, during periods of network stability, this will happen in time at most $O(\Delta)$. Thus, if an honest validator $V$ sends a vote $\mu$ to another honest validator $W$, it will take at most a constant number of rounds of communication for $W$ to verify the validity of the vote and at most a constant number of rounds to endorse it (if necessary), by which point $W$ will be ready to include $\mu$ in their next vote.
```

Comment:
As discussed, there needs to be some additional rule for sending endorsements. Otherwise if Alice cites Eve's E in her message A, and Bob cites Eve's equivocating E' in B, some validators could:
* receive A, learn about E, E', receive B, endorse B, and some might:
* receive B, learn about E, E', receive A, endorse A.

The former would never put A in their verification buffer and never endorse it. And the latter wouldn't endorse B. So it's not guaranteed that A and B will receive the required number of endorsements.